### PR TITLE
feat: adding hubot-slack types

### PR DESCRIPTION
**Pull request by O-Mutt**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
    - https://github.com/hubotio/hubot/blob/main/src/Robot.mjs#L77-L91
    - https://github.com/hubotio/hubot/blob/main/src/Adapter.mjs#L11
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

### CODE REVIEW

File: types/hubot/index.d.ts [hunk 1]
```diff
@@ -107,17 +109,23 @@ declare namespace Hubot {
...
         message: Message;
     }
 
-    class Response<A extends Adapter = Adapter, M extends Message = Message> {
-        match: RegExpMatchArray;
-        message: Message;
+    class Response<
+        A extends Adapter = Adapter,
+        M extends Message = Message,
+        R extends RegExpMatchArray | { [key: string]: string } = RegExpMatchArray
```
Comment (O-Mutt) [inline types/hubot/index.d.ts:121]: for this change the matche can be a RegExpMatchArray *OR* a custom object

https://github.com/hubotio/hubot/blob/main/src/Robot.mjs#L77-L91

File: types/hubot/index.d.ts [hunk 2]
```diff
@@ -107,17 +109,23 @@ declare namespace Hubot {
     }
 
     interface Envelope {
+        id: string;
```
Comment (O-Mutt) [inline types/hubot/index.d.ts:112]:
```suggestion
```

File: types/hubot-slack/hubot-extension.d.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+declare module 'hubot' {
+    interface Robot {
+        hearReaction(matcher: (message: Message) => boolean): void;
+        hearReaction(matcher: (message: Message) => boolean, callback: ListenerCallback<SlackBot>): void;
+        hearReaction(matcher: (message: Message) => boolean, options: {}, callback: ListenerCallback<SlackBot>): void;
+        hearMeMessage(matcher: (message: Message) => boolean, options: {}, callback: ListenerCallback<SlackBot>): void;
+        react(matcher: (message: Message) => boolean, callback: ListenerCallback<SlackBot>): void;
+        presenceChange(matcher: (message: Message) => boolean, callback: ListenerCallback<SlackBot>): void;
+        fileShared(matcher: (message: Message) => boolean, callback: ListenerCallback<SlackBot>): void;
+    }
```
Comment (O-Mutt) [inline types/hubot-slack/hubot-extension.d.ts:13]:
```suggestion
    }
    
    interface Envelope {
        id: string;
        room: string;
        user: User;
        message: Message;
    }
```

File: types/hubot/index.d.ts [hunk 4]
```diff
@@ -1,10 +1,12 @@
+/// <reference types="node" />
 import { EventEmitter } from "events";
 import { Express } from "express";
 import { Server } from "http";
 import { Options as HttpOptions, ScopedClient } from "scoped-http-client";
 
 declare namespace Hubot {
     class Adapter extends EventEmitter {
+        robot: Robot;
```
Comment (O-Mutt) [inline types/hubot/index.d.ts:9]: without this line robot isn't able to be referenced from incoming messages

https://github.com/hubotio/hubot/blob/main/src/Adapter.mjs#L11

**Comment by KeesCBakker** [comment]
Wow that is quite some work you did! I'm not sure how I can validate it. Is there a way for me to pull these changes into a project to see if this breaks something?

**Comment by KeesCBakker** [comment]
BTW... I stopped using `hubot-slack` as it is super outdated. I've started to use https://www.npmjs.com/package/@hubot-friends/hubot-slack as it is better maintained by someone that is also a maintainer for Hubot.

**Comment by O-Mutt** [comment]
> Wow that is quite some work you did! I'm not sure how I can validate it. Is there a way for me to pull these changes into a project to see if this breaks something?

You should be able to start a hubot project using typescript and configure a hear/respond matcher specific to slack.

```typescript
import { SlackBot } from 'hubot-slack';

export default function myListener(robot: Robot<SlackBot>) {
  robot.hearReaction((msg: Message) => return true, (msg: Message) => msg.respond('Hey, I heard a reaction in slack!'));
}
```

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/hubot-slack/package.json
1. types/hubot-slack/hubot-slack-tests.ts
1. types/hubot-slack/hubot-extension.d.ts
1. types/hubot-slack/index.d.ts
1. types/hubot/index.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against e580b59ec72fa489ca73fe80b6281662a0ab0eb0
</p>

File: types/hubot-slack/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+            robot_name: string,
+            robot_alias: string,
+            client: SlackClient,
+            cb: (result: unknown | SlackTextMessage) => void,
+        ): SlackTextMessage;
+    }
+
+    class SlackClient {
+        // `@` can't be used as a property in TypeScript, so we use `at` instead.
+        private atCONVERSATION_CACHE_TTL_MS: number;
```
Comment (RyanCavanaugh) [inline types/hubot-slack/index.d.ts:93]: You can write `private "@CONVERSATION_CACHE_TTL_MS": number;`

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by O-Mutt** [comment]
Ready to merge

**Pull request merged**